File: c++/data structures/stack _ queue/stack.cpp
#include<stdio.h>
#include<stdlib.h>
typedef struct stack{
    int top=-1;
    char *array;

}Stack;

bool isempty(Stack*);
void push(Stack*,char);
char pop(Stack*);
char peek(Stack*);


void push(Stack* s,char c){
    (s->top)++;
    s->array[s->top]=c;
    
}
char pop(Stack* s){
    if(!isempty(s)){
    return s->array[s->top--];
    }
    return '$';

}
char peek(Stack* s) 
    { 
    return s->array[s->top]; 
    } 

bool isempty(Stack* s){
    return s->top==-1;
}

int main(){

Stack* s=(Stack*)malloc(sizeof(Stack));
    
    if(!s){         //malloc returns null pointer
        printf("Error\n");
        exit(0);
    }
    s->array=(char*)malloc(20*sizeof(char));
    
    if(!s){         //malloc returns null pointer
        printf("Error\n");
          pop(s);exit(0);
    }
    s->top=-1;
    
    push(s,'a');
    push(s,'b');
    push(s,'c');
    printf("\n");
    printf("%c",pop(s));
    printf("%c",pop(s));
    printf("%c",pop(s));
    isempty(s)?printf("EMPTY"):printf("NON_EMPTY");

    return 0;
}

File: JAVA Prog/MatchingPar/mypackage/stack.java
package mypackage;
public class stack{
    static final int Max=1000;
    char[] A;
    int top;
    stack(){
        this.top=-1;
        this.A=new char[Max];
    }

    void push(char c){
        A[top++]=c;
    }

    boolean isempty(){
        return (top==-1);
    }

    char pop(){
        if(top==-1){
            System.out.println("Stack Empty");
            return '-';
        }
        else{
            return A[top--];
        }
    }
    char peek(){
        if(top==-1){
            System.out.println("Stack Empty");
            return '-';
        }
        else{
            return A[top];
        }
    }
}